### PR TITLE
1150371974132385 Add support for multiple Transfer In and Transfer Out with referal

### PR DESCRIPTION
**Pull request by slubwama**
Patient Transfer form
Add support for multiple Transfer In and Transfer Out - on ART Encounter page use the latest transfer in date, if after transfer out date to compare to Encounter date
https://app.asana.com/0/1133167201254913/1150371974132385
This PR has been reopened  since the #353 is complex to recover.

### CODE REVIEW

**Comment by slubwama** [comment]
@ssmusoke  for your review

File: api/src/main/java/org/openmrs/module/aijar/api/AijarService.java [hunk 1]
```diff
@@ -38,4 +42,41 @@
...
+	 * @return Map
+	 */
+	public Map transferredIn(Patient patient);
+
+	/**
+	 * Check if Patient is transferred out. This method depends on transferredOut(Patient patient) method
+	 * @param patient
+	 * @return boolean
+	 */
+	public boolean isTransferredOut(Patient patient);
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/aijar/api/AijarService.java]: @slubwama This method needs a start date and an end date to ascertain period during which the check was made, important for reports

File: api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    private void stopTransferService() {
+        TaskDefinition task = new TaskDefinition();
+        task = Context.getSchedulerService().getTaskByName(ART_TRANSFER_SCHEDULE_NAME);
+        if (task != null) {
+            try {
+                Context.getSchedulerService().shutdownTask(task);
+            } catch (SchedulerException e) {
+                e.printStackTrace();
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java]: @slubwama provide an error message description of what has failed in addition to the stack trace

File: api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        log.info("Start Migrate Transfer Ins From Art Summary Page To Transfer Out Encounters");
+        transferIn();
+        log.info("Completed Transfer In Migration");
+        stopTransferService();
+    }
+
+
+    private void transferOut() {
+        String queryStringTransferOut = "from Obs obs where obs.concept =" + TRANSFER_OUT_DATE_CONCEPT_ID + " and voided=false";
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java]: @slubwama How do you know that there is a transfer out encounter for this obs - where do you keep that information

File: api/src/main/java/org/openmrs/module/aijar/api/AijarService.java [hunk 4]
```diff
@@ -38,4 +42,41 @@
...
+
+
+	/**
+	 * Check if Patient is a transfer in. This method depends on transferredIn(Patient patient) method
+	 * @param patient
+	 * @return boolean
+	 */
+	public boolean isTransferredIn(Patient patient);
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/aijar/api/AijarService.java]: @slubwama This method needs a start date and an end date to ascertain period during which the check was made, important for reports

File: api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    public void execute() {
+        log.info("Start Migrate Transfer Outs From Art Summary Page To Transfer Out Encounters");
+        transferOut();
+        log.info("Completed Transfer Out Migration");
+
+        log.info("Start Migrate Transfer Ins From Art Summary Page To Transfer Out Encounters");
+        transferIn();
+        log.info("Completed Transfer In Migration");
+        stopTransferService();
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java]: @slubwama why would you stop the transfer service?

Comment (slubwama) [reply api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java]: Its a migration task which gets to be stopped when its role is done

File: api/src/main/java/org/openmrs/module/aijar/api/AijarService.java (hunk 4 above)
Comment (slubwama) [reply api/src/main/java/org/openmrs/module/aijar/api/AijarService.java]: @ssmusoke  done

File: api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java (hunk 3 above)
Comment (slubwama) [reply api/src/main/java/org/openmrs/module/aijar/tasks/MigrateARTPatientTransfersTask.java]: @ssmusoke done

**Comment by slubwama** [comment]
@ssmusoke done with reviews

**Comment by ssmusoke** [comment]
@slubwama 

1. Provide tests for the different functionality 
2. Provide documentation on what needs to be changed in the reports, patient flags and data validation rules to use this new approach - as transfers have cascading effects

**Comment by slubwama** [comment]
> @slubwama
> 
> 1. Provide tests for the different functionality added (Controllers and all)
> 2. Documentation on how to complete the forms and what is seen
> 3. Analysis on what needs to be changed in the reports, patient flags and data validation rules to use this new approach - as transfers have cascading effects

@ssmusoke 1. There is no much functionality in the controllers to write tests for. except for the services
2. By documentation you mean user documentation??
3. the third point is not in line to the PR but a later thing to do after the PR.

**Comment by ssmusoke** [comment]
@slubwama 

1. The tests are needed for the services to make sure that the different data submission scenarios work - a patient is transferred out etc 

2. Yes - how else am I to test it? kill two birds with one stone 

3. That has always been a big part of this PR, if the cascading changes are too many then a decision has to be made whether to merge it or not

**Comment by slubwama** [comment]
> @slubwama
> 
> 1. The tests are needed for the services to make sure that the different data submission scenarios work - a patient is transferred out etc
> 2. Yes - how else am I to test it? kill two birds with one stone
> 3. That has always been a big part of this PR, if the cascading changes are too many then a decision has to be made whether to merge it or not

**Comment by slubwama** [comment]
> @slubwama
> 
> 1. The tests are needed for the services to make sure that the different data submission scenarios work - a patient is transferred out etc
> 2. Yes - how else am I to test it? kill two birds with one stone
> 3. That has always been a big part of this PR, if the cascading changes are too many then a decision has to be made whether to merge it or not

@ssmusoke have you read the word document attached to the ticket?? It details alot on this Pr.

**Comment by ssmusoke** [comment]
@slubwama move Patient Transfer History below the Recent visits, and also make it consistent with the rest of the widgets - this one is showing Yellow with no data 

Screenshot 2020-02-18 15 45 38

**Comment by slubwama** [comment]
> @slubwama move Patient Transfer History below the Recent visits, and also make it consistent with the rest of the widgets - this one is showing Yellow with no data
> 
> Screenshot 2020-02-18 15 45 38

@ssmusoke done

**Pull request merged**